### PR TITLE
xrEngine: allow to run without game module

### PR DESCRIPTION
**Pull request by vTurbine**
Also do not crash if there are no levels data

### CODE REVIEW

File: src/xrEngine/EngineAPI.cpp [hunk 1]
```diff
@@ -130,13 +130,18 @@ void CEngineAPI::Initialize(void)
     InitializeRenderers();
 
     hGame = XRay::LoadModule("xrGame");
-    R_ASSERT2(hGame->IsLoaded(), "Game DLL raised exception during loading or there is no game DLL at all");
-
-    pCreate = (Factory_Create*)hGame->GetProcAddress("xrFactory_Create");
-    R_ASSERT(pCreate);
+    if (!hGame->IsLoaded())
+    {
+        Msg("! Game DLL raised exception during loading or there is no game DLL at all");
```
Comment (Xottab-DUTY) [inline src/xrEngine/EngineAPI.cpp]: This is not a good thing to allow xrGame loading to fail by default, consider adding `-no_game` key which will just disable xrGame loading.

File: src/xrEngine/EngineAPI.h [hunk 2]
```diff
@@ -83,9 +83,9 @@ class ENGINE_API CEngineAPI
 
 ENGINE_API bool is_enough_address_space_available();
 
-#define NEW_INSTANCE(a) Engine.External.pCreate(a)
+#define NEW_INSTANCE(a) Engine.External.pCreate ? Engine.External.pCreate(a) : nullptr
```
Comment (Xottab-DUTY) [inline src/xrEngine/EngineAPI.h]: It's better and more safe to assign pointer to a dummy function:
- create dummy will just return nullptr 
- destroy dummy will assert that passed pointer is nullptr.

File: src/xrEngine/x_ray.cpp [hunk 3]
```diff
@@ -393,7 +393,11 @@ void CApplication::Level_Scan()
     Levels.clear();
 
     xr_vector<char*>* folder = FS.file_list_open("$game_levels$", FS_ListFolders | FS_RootOnly);
-    //. R_ASSERT (folder&&folder->size());
+    if (!folder)
+    {
+        Msg("! No levels found in game data");
```
Comment (Xottab-DUTY) [inline src/xrEngine/x_ray.cpp]: There's no need to call Msg if we don't need to format log.
```suggestion
        Log("! No levels found in game data");
```

File: src/xrEngine/x_ray.cpp [hunk 4]
```diff
@@ -182,7 +182,6 @@ void CApplication::OnEvent(EVENT E, u64 P1, u64 P2)
         LPSTR op_client = LPSTR(P2);
         Level_Current = u32(-1);
         R_ASSERT(nullptr == g_pGameLevel);
-        R_ASSERT(nullptr != g_pGamePersistent);
```
Comment (Xottab-DUTY) [inline src/xrEngine/x_ray.cpp]: This assert should be here.

File: src/xrEngine/main.cpp [hunk 5]
```diff
@@ -324,7 +324,6 @@ ENGINE_API void Startup()
     Device.WaitEvent(spatialCreated);
 
     g_pGamePersistent = dynamic_cast<IGame_Persistent*>(NEW_INSTANCE(CLSID_GAME_PERSISTANT));
-    R_ASSERT(g_pGamePersistent);
```
Comment (Xottab-DUTY) [inline src/xrEngine/main.cpp:327]: Here should be something like this:
```cpp
R_ASSERT(g_pGamePersistent || Engine.External.IsNoGame());
```

File: src/xrEngine/x_ray.cpp [hunk 6]
```diff
@@ -232,7 +233,6 @@ void CApplication::OnEvent(EVENT E, u64 P1, u64 P2)
         LPSTR demo_file = LPSTR(P1);
 
         R_ASSERT(nullptr == g_pGameLevel);
-        R_ASSERT(nullptr != g_pGamePersistent);
```
Comment (Xottab-DUTY) [inline src/xrEngine/x_ray.cpp]: This should be here too. I'm not sure how demo can be started without xrGame 🙂

**Review by Xottab-DUTY — CHANGES_REQUESTED** [review]

File: src/xrEngine/EngineAPI.cpp [hunk 7]
```diff
@@ -165,8 +172,8 @@ void CEngineAPI::Destroy(void)
 {
     hGame = nullptr;
     hTuner = nullptr;
-    pCreate = nullptr;
-    pDestroy = nullptr;
+    factoryCreate  = dummyCreate;
+    factoryDestroy = dummyDestroy;
```
Comment (Xottab-DUTY) [inline src/xrEngine/EngineAPI.cpp]: They should be nullptr.

File: src/xrEngine/EngineAPI.cpp [hunk 8]
```diff
@@ -39,13 +39,15 @@ static bool r2_available = false;
 //////////////////////////////////////////////////////////////////////
 
 void __cdecl dummy(void) {}
+constexpr auto dummyCreate  = [](auto){ return nullptr; };
+constexpr auto dummyDestroy = [](auto){ };
```
Comment (Xottab-DUTY) [inline src/xrEngine/EngineAPI.cpp]: It's better to make it plain functions. Also, destroy should assert that passed pointer is nullptr.

File: src/xrEngine/EngineAPI.h [hunk 9]
```diff
@@ -60,8 +62,8 @@ class ENGINE_API CEngineAPI
 public:
     BENCH_SEC_SCRAMBLEMEMBER1
 
-    Factory_Create*  pCreate;
-    Factory_Destroy* pDestroy;
+    std::function<Factory_Create>  factoryCreate;
+    std::function<Factory_Destroy> factoryDestroy;
```
Comment (Xottab-DUTY) [inline src/xrEngine/EngineAPI.h]: What's the point on using std::function? My argument against using it – is that it is slower than plain pointer.

File: src/xrEngine/EngineAPI.h [hunk 10]
```diff
@@ -4,13 +4,15 @@
...
 #pragma once
-#include <memory>
 
 #include "xrEngine/Engine.h"
 #include "xrCore/ModuleLookup.hpp"
 #include "xrCore/clsid.h"
 #include "xrCore/xrCore_benchmark_macros.h"
 
+#include <functional>
+#include <memory>
```
Comment (Xottab-DUTY) [inline src/xrEngine/EngineAPI.h:13]: Saying just in case: this change is good, don't move <memory> back to top.

File: src/xrEngine/EngineAPI.h [hunk 11]
```diff
@@ -76,16 +78,17 @@ class ENGINE_API CEngineAPI
     void Destroy();
 
     void CreateRendererList();
+    bool SkipGameModuleLoading() const { return !!strstr(Core.Params, "-nogame"); }
```
Comment (Xottab-DUTY) [inline src/xrEngine/EngineAPI.h]: Just to be explicit =))
```suggestion
    bool CanSkipGameModuleLoading() const { return !!strstr(Core.Params, "-nogame"); }
```

File: src/xrEngine/main.cpp [hunk 12]
```diff
@@ -324,7 +324,7 @@ ENGINE_API void Startup()
...
+    R_ASSERT(g_pGamePersistent || Engine.External.SkipGameModuleLoading());
```
Comment (Xottab-DUTY) [inline src/xrEngine/main.cpp]: Just to be explicit =))
```suggestion
    R_ASSERT(g_pGamePersistent || Engine.External.CanSkipGameModuleLoading());
```

**Review by Xottab-DUTY — CHANGES_REQUESTED** [review]

File: src/xrEngine/EngineAPI.cpp (hunk 8 above)
Comment (vTurbine) [reply src/xrEngine/EngineAPI.cpp]: Nope. This will crash when trying to release pGamePersistent with `-nogame` option

**Review by Xottab-DUTY — APPROVED** [review]
Thank you for your patience!

**Pull request merged**